overlord/devicestate/users.go

```diff
@@ -0,0 +1,359 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+/*
+ * Copyright (C) 2022 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package devicestate
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os/user"
+	"path/filepath"
+	"time"
+
+	"github.com/snapcore/snapd/asserts"
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/overlord/assertstate"
+	"github.com/snapcore/snapd/overlord/auth"
+	"github.com/snapcore/snapd/overlord/snapstate"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/release"
+	"github.com/snapcore/snapd/strutil"
+)
+
+var (
+	osutilAddUser = osutil.AddUser
+	osutilDelUser = osutil.DelUser
+	userLookup    = user.Lookup
+)
+
+// UserError is used for errors which are known to snapd. This is errors
+// that originate from this code here.
+type UserError struct {
+	Err error
+}
+
+func (e *UserError) Error() string {
+	return e.Err.Error()
+}
+
+// CreatedUser is the returned result which wraps some resulting data
+// from a create user operation.
```

Comment:
nitpick: `returned reuslt which wraps some resulting` sounds odd, maybe something like this?
```suggestion
// CreatedUser carries a username and SSH keys resulting from the user creation.
```

Reply:
I agree the comment could maybe be better, but I don't think it's neccessary for the comment to iterate the members. I was thinking something more general for this comment, just explaining that it contains resulting data from a create user operation, as which data it contains is fairly obvious (or there should be per-data member comments if a member is not obvious)